### PR TITLE
[SPARK-30818][SPARKR][ML] Add SparkR LinearRegression wrapper

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This pull request adds SparkR wrapper for `LinearRegression`

- Supporting `org.apache.spark.ml.rLinearRegressionWrapper`.
- `LinearRegressionModel` S4 class.
- Corresponding `spark.lm` predict, summary and write.ml generics.
- Corresponding docs and tests.

### Why are the changes needed?

Feature parity.

### Does this PR introduce any user-facing change?

No (new API).

### How was this patch tested?

New unit tests.

### CODE REVIEW

File: R/pkg/R/mllib_regression.R [hunk 1]
```diff
@@ -540,3 +546,150 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' predictions <- predict(model, df)
+#'
+#' # save and load the model
+#' path <- "path/to/model"
+#' write.ml(model, path)
+#' savedModel <- read.ml(path)
+#' summary(savedModel)
+#' }
+#' @note spark.lm since 3.1.0
+setMethod("spark.lm", signature(data = "SparkDataFrame", formula = "formula"),
```
Comment (zero323) [inline R/pkg/R/mllib_regression.R]: It is for discussion what name should be used here. Long name (`spark.linearRegression`) might be preferred, but we already have `spark.glm` so this seems to be justified.

`spark.glm` also exports `glm`, but I think it is confusing, and inconsistent,  so I've decided not to add `lm`.

File: R/pkg/R/mllib_regression.R [hunk 2]
```diff
@@ -540,3 +546,147 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+          function(data, formula,
+                   maxIter = 100L, regParam = 0.0, elasticNetParam = 0.0,
+                   tol = 1e-6, standardization = TRUE,
+                   solver = c("auto", "l-bfgs", "normal"),
+                   weightCol = NULL, aggregationDepth = 2L,
+                   loss = c("squaredError", "huber"), epsilon = 1.35,
+                   stringIndexerOrderType = c("frequencyDesc", "frequencyAsc",
+                                              "alphabetDesc", "alphabetAsc")) {
+
+
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: nit: delete extra blank line? There are several other places that have two blank lines as well.

**Comment by zero323** [comment]
Retest this please.

**Comment by zero323** [comment]
Retest this please.

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+import org.apache.spark.ml.feature.RFormula
+import org.apache.spark.ml.r.RWrapperUtils._
+import org.apache.spark.ml.regression.{LinearRegression, LinearRegressionModel}
+import org.apache.spark.ml.util._
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+private[r] class LinearRegressionWrapper private (
+    val pipeline: PipelineModel,
+    val features: Array[String]) extends MLWritable {
+  import LinearRegressionWrapper._
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: Nit: seems the import is not needed?

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  lazy val numFeatures: Int = linearRegressionModel.numFeatures
+
+  def transform(dataset: Dataset[_]): DataFrame = {
+    pipeline.transform(dataset)
+      .drop(linearRegressionModel.getFeaturesCol)
+  }
+
+
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: super nit: remove extra blank line

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      .setAggregationDepth(aggregationDepth)
+      .setLoss(loss)
+      .setEpsilon(epsilon)
+      .setFeaturesCol(rFormula.getFeaturesCol)
+
+    if (weightCol != null) {
+      lm.setWeightCol(weightCol)
+    }
+
+
```
Comment (huaxingao) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: super nit: remove extra blank line

File: R/pkg/R/mllib_regression.R [hunk 6]
```diff
@@ -540,3 +546,145 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' @param maxIter maximum iteration number.
+#' @param regParam the regularization parameter.
+#' @param elasticNetParam the ElasticNet mixing parameter, in range [0, 1].
+#'        For alpha = 0, the penalty is an L2 penalty. For alpha = 1, it is an L1 penalty.
+#' @param tol convergence tolerance of iterations.
+#' @param standardization whether to standardize the training features before fitting the model.
+#' @param weightCol weight column name.
+#' @param aggregationDepth suggested depth for treeAggregate (>= 2).
+#' @param loss the loss function to be optimized. Supported options: "squaredError" and "huber"
+#' @param epsilon the shape parameter to control the amount of robustnes
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: super nit: ```robustnes``` -> ```robustness```

File: R/pkg/R/mllib_regression.R [hunk 7]
```diff
@@ -540,3 +546,145 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' @name spark.lm
+#' @seealso \link{read.ml}
+#' @examples
+#' \dontrun{
+#' df <- read.df("data/mllib/sample_linear_regression_data.txt", source = "libsvm")
+#'
+#' # fit Linear Regression Model
+#' model <- spark.lm(
+#'            df, label ~ features,
+#'            regParam = 0.01, maxIter = 10, fitLinear = TRUE
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: Seems ```fitLinear``` is not a param for lm?
Also the indentation of these two lines don't seem to be quite right.

File: R/pkg/R/mllib_regression.R [hunk 8]
```diff
@@ -540,3 +546,145 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+                                "fit",
+                                data@sdf,
+                                formula,
+                                as.integer(maxIter),
+                                as.numeric(regParam),
+                                as.numeric(elasticNetParam),
+                                as.numeric(tol),
+                                as.logical(standardization),
+                                solver,
+                                weightCol,
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: Do the following for ```weightCol```?
```
            if (!is.null(weightCol) && weightCol == "") {
              weightCol <- NULL
            } else if (!is.null(weightCol)) {
              weightCol <- as.character(weightCol)
            }
```

File: R/pkg/R/mllib_regression.R [hunk 9]
```diff
@@ -540,3 +546,145 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
           function(object, path, overwrite = FALSE) {
             write_internal(object, path, overwrite)
           })
+
+#' Linear Regression Model
+#'
+#' \code{spark.lm} fits a linear regression model against a SparkDataFrame.
+#' Users can call \code{predict} to make
+#' predictions on new data, and \code{write.ml}/\code{read.ml} to save/load fitted models.
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: nit: add description for ```summary``` too?

File: examples/src/main/r/ml/lm_with_elastic_net.R [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+# Initialize SparkSession
+sparkR.session(appName = "SparkR-ML-lm-example")
+
+# $example on$
+# Load training data
+df <- read.df("data/mllib/sample_linear_regression_data.txt", source = "libsvm")
+training <- df
+test <- df
+
+
```
Comment (huaxingao) [inline examples/src/main/r/ml/lm_with_elastic_net.R]: super nit: remove extra blank line?

File: R/pkg/R/mllib_regression.R [hunk 11]
```diff
@@ -540,3 +546,145 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+            list(
+              coefficients = coefficients,
+              numFeatures = numFeatures
+            )
+          })
+
+#  Predicted values based on an LinearRegressionModel model
+
+#' @param newData a SparkDataFrame for testing.
+#' @return \code{predict} returns the predicted values based on an LinearRegressionModel.
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R]: super nit: ```an``` -> ```a```

**Comment by huaxingao** [comment]
also cc @felixcheung

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    val pipeline = new Pipeline()
+      .setStages(Array(rFormulaModel, lm))
+      .fit(data)
+
+    new LinearRegressionWrapper(pipeline, features)
+  }
+
+  override def read: MLReader[LinearRegressionWrapper] = new LinearRegressionWrapperReader
+
+  override def load(path: String): LinearRegressionWrapper = super.load(path)
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: Is this redundant?

**Review by viirya — COMMENTED** [review]
Seems fine.

**Comment by srowen** [comment]
@zero323 do you have any thoughts on the last few comments there? Seems OK. I'm actually surprised this didn't exist before!

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala (hunk 3 above)
Comment (zero323) [reply mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: Indeed, thanks.

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala (hunk 12 above)
Comment (zero323) [reply mllib/src/main/scala/org/apache/spark/ml/r/LinearRegressionWrapper.scala]: Yes, that seems to be the case.

**Comment by zero323** [comment]
> @zero323 do you have any thoughts on the last few comments there? Seems OK. I'm actually surprised this didn't exist before!

Sorry, busy days. I think the problem is the amount of boilerplate we need to implement any R wrapper. I've been meaning To address that, one problem at the time, maybe starting with resurrecting [SPARK-20729](https://github.com/apache/spark/pull/17969).

File: R/pkg/R/mllib_regression.R [hunk 13]
```diff
@@ -540,3 +546,149 @@ setMethod("write.ml", signature(object = "AFTSurvivalRegressionModel", path = "c
...
+#' @param stringIndexerOrderType how to order categories of a string feature column. This is used to
+#'                               decide the base level of a string feature as the last category
+#'                               after ordering is dropped when encoding strings. Supported options
+#'                               are "frequencyDesc", "frequencyAsc", "alphabetDesc", and
+#'                               "alphabetAsc". The default value is "frequencyDesc". When the
+#'                               ordering is set to "alphabetDesc", this drops the same category
+#'                               as R when encoding strings.
+#' @param ... additional arguments passed to the method.
+#' @return \code{spark.lm} returns a fitted Linear Regression Model.
+#'
```
Comment (huaxingao) [inline R/pkg/R/mllib_regression.R:581]: @HyukjinKwon Is this blank line OK? I remember in one of my PR, Felix said the blank line is significant in roxygen2 and asked me to remove.

**Comment by srowen** [comment]
@zero323 this one needs a rebase too.
I also might remove that blank line per comments above.

File: R/pkg/R/mllib_regression.R (hunk 13 above)
Comment (zero323) [reply R/pkg/R/mllib_regression.R:581]: In general blank lines are acceptable and even recommended after long block (special case is a header of the comment, where it is meaningful. But I am fine with removing it.

**Comment by zero323** [comment]
Retest this please.

**Comment by huaxingao** [comment]
retest this please

**Comment by zero323** [comment]
> @zero323 this one needs a rebase too.
> I also might remove that blank line per comments above.

Done @srowen

**Comment by srowen** [comment]
Merged to master

**Comment by zero323** [comment]
Thanks everyone!

**Pull request closed without merging**